### PR TITLE
Build gitlabci file when accepted to main branch.

### PR DESCRIPTION
**Pull request by mvr320**
If we create a protected environment we can even push to another registry for the other containers without giving away the token.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -1,34 +1,43 @@
...
+  only:
+    changes:
+      - docker-gitlabci/*
   script:
-    - ./docker-gitlabci/build.sh
+    - cd docker-gitlabci
+    - ./build.sh $GITLABCI_VERSION
 
 build-gitlabci:
+  environment: maintainer-approved
```
Comment (nickygerritsen) [inline .gitlab-ci.yml:26]: What does this do?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
     changes:
       - docker-gitlabci/*
   script:
-    - ./docker-gitlabci/build.sh
+    - cd docker-gitlabci
+    - ./build.sh $GITLABCI_VERSION
     - docker login -u $CI_REGISTRY_USER -p $CI_REGISTRY_PASSWORD $CI_REGISTRY
     - docker push $GITLABCI_IMAGE_TEST
 
 image-gitlabci:
```
Comment (nickygerritsen) [inline .gitlab-ci.yml:39]: Do we want another stage if we only do it on master anyway? Why not push directly to the correct image?

**Review by nickygerritsen — COMMENTED** [review]
I like this. For the contributor image we can also push it to gitlab I guess. No need to have it on docker hub if we update the compose file.

For the main images this is a bit harder since it requires a zip to be present on the website with the release. We could make a manual step (or multiple) and ask for the version to build.

File: .gitlab-ci.yml (hunk 1 above)
Comment (mvr320) [reply .gitlab-ci.yml:26]: The idea is that you add your special token for dockerhub in gitlab, and only runs triggered by maintainers in gitlab can trigger runs for that environment. In this way an external contributor cannot steal your tokens...

Because when that works, we can than also have that you can trigger the uploading either by merging after the zip is there or by using the "run pipeline" where you add a variable for where the zip can be downloaded from.

TL;DR not needed for this one, but adds some form of obfuscation in such a way that someone cannot push a new gitlabci which isn't tested yet.

File: .gitlab-ci.yml (hunk 2 above)
Comment (mvr320) [reply .gitlab-ci.yml:39]: Thats a good question...

I think its to meta to use the tests from domjudge:.gitlab-ci.yml here to test if you want this updated image, so I'll just remove the 2nd stage.

**Comment by mvr320** [comment]
> 
> 
> I like this. For the contributor image we can also push it to gitlab I guess. No need to have it on docker hub if we update the compose file.
> 
> For the main images this is a bit harder since it requires a zip to be present on the website with the release. We could make a manual step (or multiple) and ask for the version to build.

you can run:
https://gitlab.com/mvasseur/domjudge-packaging/-/pipelines/new

ofcourse the mvasseur might not be the best project to run this from.

**Comment by mvr320** [comment]
@nickygerritsen for the variables I might need some help from you or would like you to set this up (as you've done it before)

Do you think you want the idea to also build via gitlab? (so with the protected environments), as I think I can incorporate it, but if its not wanted I rather not fill the template with unused jobs.

**Comment by nickygerritsen** [comment]
> @nickygerritsen for the variables I might need some help from you or would like you to set this up (as you've done it before)
> 
> Do you think you want the idea to also build via gitlab? (so with the protected environments), as I think I can incorporate it, but if its not wanted I rather not fill the template with unused jobs.

Currently I am always the one building the docker images so if we can automate it it is less work for me. Less work = more better

**Comment by mvr320** [comment]
> Currently I am always the one building the docker images so if we can automate it it is less work for me. Less work = more better

See f17da2d for a possible implementation, we should check if it works when DOMjudge 8.0 is released. But if this indeed works as how I think it works you don't have to be the only one anymore who can release... although its probably with your DOMjudge credentials.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      - master
+    changes:
+      - docker-contributor/*
+  script:
+    - cd docker-contributor
+    - docker build -t $CONTRIBUTOR_IMAGE
+    - docker push $CONTRIBUTOR_IMAGE
+
+release-DOMjudge:
+  <<: *registry_dockerhub
```
Comment (nickygerritsen) [inline .gitlab-ci.yml:77]: This needs some `manual` keyword right?

File: .gitlab-ci.yml (hunk 1 above)
Comment (nickygerritsen) [reply .gitlab-ci.yml:39]: I thought we could do this with a protected environment variable in GitLab? That we only expose it on master.
I think setting an `environment` key itself doesn't do much? Or do you have an example of what it does?

Comment (mvr320) [reply .gitlab-ci.yml:39]: You need both,

https://docs.gitlab.com/ee/ci/environments/protected_environments.html

Currently we do not have an environment yet, (I cannot select it) so we need to add an environment in the tasks, (although I think it can be global instead of what I do now) which will than only expose the variables of username/password when someone with maintainer role (or higher) in gitlab runs this (current list: https://gitlab.com/groups/DOMjudge/-/group_members).

As we discussed to actualy make this a seperate pipeline I can shift this a bit, if you change my access to developer we can even set an environment for gitlabci deployment (allowing me and Mart to also do this) and set another environment for DOMjudge release deployment (allowing this only to maintainers -> All the owners)

TL;DR for the gitlab interface we atleast need 1 environment (there is not a default) and environment !== branch, and you're right, I also need to setup additional things in GitLab

Comment (nickygerritsen) [reply .gitlab-ci.yml:39]: I was thinking to use this:
image
i.e. the 'Protect variables' thing and then make only `master` a protected branch. Then we do not need the environment I think?

Comment (mvr320) [reply .gitlab-ci.yml:39]: I think you're right, I'll try to build something in mvasseur for gitlabci tonight.
